### PR TITLE
[deepspeed] partial ZeRO-3 support

### PR DESCRIPTION
**Pull request by stas00**
Fixes: https://github.com/huggingface/diffusers/issues/1865

Here is what's happening here:

When Deepspeed ZeRO-3 is used in `examples/text_to_image/train_text_to_image.py` the 3 models (`text_encoder`, `vae`, `unet`) get partitioned with `zero.Init` but only `unet` gets `accelerate.prepare` so Deepspeed doesn't know to automatically gather the already partitioned weights before each `forward` call for `text_encoder` and `vae`. That's the source of the problem. 

Now, to provide full ZeRO-3 support you have to contact https://github.com/huggingface/accelerate and ask them to extend support to this situation. Specifically:

Currently Accelerate doesn't know how to handle multiple models.

For this to work properly all models must be run through `accelerate.prepare`. But Accelerate will try to assign the same optimizer with the same weights to all models during `deepspeed.initialize`, which of course doesn't work. It needs to assign the correct weights to the correct optimizer, so that the param_groups are populated with the weights that are inside the model itself. In the case of frozen models there should be no optimizer at all (or `param_groups` should be empty).

It will need to be able to support something like the following in `accelerate.prepare`:
```
# frozen models
deepspeed.initialize(model=text_encoder, model_parameters=None, optimizer=None, ...)
deepspeed.initialize(model=vae, model_parameters=None, optimizer=None, ...)
# normal model
deepspeed.initialize(model=unet, model_parameters=unet_model_parameters, optimizer=optimizer, ...)
```

For now this PR will partially support Deepspeed ZeRO-3, by excluding the 2 frozen models from being partitioned during zero.Init which gets called during `from_pretrained`. This is done for `examples/text_to_image/train_text_to_image.py` only, but can be replicated to other scripts if needed.

So `CLIPTextModel` and `AutoencoderKL` will not enjoy the parameter sharding across multiple gpus and only `UNet2DConditionModel` will get zero sharded, and so memory savings will only apply to the latter. The previous 2 will have full copies on each gpu.

----------------

to reviewers: As I'm not familiar with this project's structure I will leave to maintainers to choose where to put the helper code inside the example I modified to work. I just put it in one place for visibility.

----------------------

If you want full ZeRO-3 support for optimal training efficiency, here is what you need to do

1. resolve the issue with Accelerate as discussed in this OP
2. then will have to ask Deepspeed to provide support for multiple models - but we can't do that until stage 1 is done.

### CODE REVIEW

**Comment by stas00** [comment]
the ci failures look to be unrelated - some timeout.

**Comment by uygnef** [comment]
Thank a lot. 

> Fixes: #1865
> 
> Here is what's happening here:
> 
> When Deepspeed ZeRO-3 is used in `examples/text_to_image/train_text_to_image.py` the 3 models (`text_encoder`, `vae`, `unet`) get partitioned with `zero.Init` but only `unet` gets `accelerate.prepare` so Deepspeed doesn't know to automatically gather the already partitioned weights before each `forward` call for `text_encoder` and `vae`. That's the source of the problem.
> 
> Now, to provide full ZeRO-3 support you have to contact https://github.com/huggingface/accelerate and ask them to extend support to this situation. Specifically:
> 
> Currently Accelerate doesn't know how to handle multiple models.
> 
> For this to work properly all models must be run through `accelerate.prepare`. But Accelerate will try to assign the same optimizer with the same weights to all models during `deepspeed.initialize`, which of course doesn't work. It needs to assign the correct weights to the correct optimizer, so that the param_groups are populated with the weights that are inside the model itself. In the case of frozen models there should be no optimizer at all (or `param_groups` should be empty).
> 
> It will need to be able to support something like the following in `accelerate.prepare`:
> 
> ```
> # frozen models
> deepspeed.initialize(model=text_encoder, model_parameters=None, optimizer=None, ...)
> deepspeed.initialize(model=vae, model_parameters=None, optimizer=None, ...)
> # normal model
> deepspeed.initialize(model=unet, model_parameters=unet_model_parameters, optimizer=optimizer, ...)
> ```
> 
> For now this PR will partially support Deepspeed ZeRO-3, by excluding the 2 frozen models from being partitioned during zero.Init which gets called during `from_pretrained`. This is done for `examples/text_to_image/train_text_to_image.py` only, but can be replicated to other scripts if needed.
> 
> So `CLIPTextModel` and `AutoencoderKL` will not enjoy the parameter sharding across multiple gpus and only `UNet2DConditionModel` will get zero sharded, and so memory savings will only apply to the latter. The previous 2 will have full copies on each gpu.
> 
> to reviewers: As I'm not familiar with this project's structure I will leave to maintainers to choose where to put the helper code inside the example I modified to work. I just put it in one place for visibility.
> 
> If you want full ZeRO-3 support for optimal training efficiency, here is what you need to do
> 
> 1. resolve the issue with Accelerate as discussed in this OP
> 2. then will have to ask Deepspeed to provide support for multiple models - but we can't do that until stage 1 is done.

Thanks a lot. It works.
"then will have to ask Deepspeed to provide support for multiple models - but we can't do that until stage 1 is done." If stage 1 for multiple models hasn't been completed yet, how does DeepSpeed-RLHF use zero-optimal?

**Comment by stas00** [comment]
Last I checked it had troubles with multiple models, but perhaps it was in a particular situation, and it might already be working. So if you'd like to take a lead on that, please file a feature request at accelerate to solve what I described in the OP. And once this is done we can then work on the deepspeed side.

**Comment by stas00** [comment]
ok, so I was just talking to @ and he shared that the newly released DeepSpeed-Chat https://github.com/microsoft/DeepSpeedExamples/tree/master/applications/DeepSpeed-Chat which coincided with Deepspeed 0.9.0 release now support multiple DS engines, e.g. see:

python /hf/Megatron-DeepSpeed-master/tools/convert_checkpoint/inspect_checkpoint.py /hf/m4-master-3/save_dir/opt_step-10/accelerator_state/pytorch_model/zero_pp_rank_0_mp_rank_00_optim_states.pt

https://github.com/microsoft/DeepSpeedExamples/blob/d22b82b7d167f3e432234795da7fdf7965db8891/applications/DeepSpeed-Chat/training/step3_rlhf_finetuning/rlhf_engine.py#L145-L146

Therefore, let's ping @pacman100 - Sourab, please kindly have a look at this OP, we need to have accelerate support multiple DS engines, each with a different setup and optimizer, to make it work with diffusion models, which have at least 3 independent models.

**Comment by uygnef** [comment]
> 

hi @stas00 
 I have published the feature request at https://github.com/huggingface/diffusers/issues/3098.

**Comment by pacman100** [comment]
Hello @stas00,

The thing is that there are multiple challenges with this which need discussion:
1. When there are multiple models, and if one of the models required zero-3 while other require zero-2 (for example actor model requires zero-3 but the small reward model needs zero-2). it will activate zero-init when remaining don't require it and it will involve a lot of juggling to use the disabling context manager as you did above ...
2. Accelerate currently takes as input via `accelerate config` questionnaire only single set of config, but different models will require different config setting. In the above DeepSpeedChat they are having helper functions to get different ds_config for each of the components such as actor and critic models. Accelerate can't have helper functions just for this usage as it won't be generic.
3.  Even if the multiple config is somehow supported, big question about `auto` values and how to fill them, if user passes a few of the values via cmd args, then which of the multiple configs will it belong to? And then the remaining values being filled with `prepare` call, different models might need different values for gradient accumulation, train batch size per gpu ... So, how would `prepare` work in that case as it can't distinguish this dataloader is meant for this model and not that model ...

**Comment by tjruwase** [comment]
@pacman100, these are great questions that you raise and essentially the same ones we wrestled with in the DeepSpeed Chat project. I would like to think that our solutions are general, but they did require a re-think of how DeepSpeed worked. 

For example, for (1), we do support actor and critic models using different zero stages as shown in our 1.3b actor on single GPU [example](https://github.com/microsoft/DeepSpeedExamples/blob/master/applications/DeepSpeed-Chat/training/step3_rlhf_finetuning/training_scripts/single_gpu/run_1.3b.sh). We also observed that juggling context managers could be simplified with some [code refactoring](https://github.com/microsoft/DeepSpeedExamples/blob/cd19b3bf1e5b60dd73b09c7463da4eedada1eed7/applications/DeepSpeed-Chat/training/utils/model/model_utils.py#L18). 

If I sound a bit bullish, it is not because I want to diminish these challenges, but rather because I am excited to be tackling these difficult yet unavoidable problems to support the next generation of ML models.

**Comment by stas00** [comment]
Sourab, I'm sure there are even more issues than what you listed. 

When we did the initial integration in HF Trainer and when you forked it to Accelerate there was always one model - as DS didn't support anything else. But now it does so what I propose is that we design a new solution that accommodates multiple models each potentially having its own needs.

Wrt `accelerate config` - you already know my stand on it - it's a great approach for a very simplistic beginner get started quickly way, but it doesn't work for advanced uses. Advanced uses require custom config files for each unique situation and relying on a single global config file makes no sense. 

Bottom line, do you want to meet and brainstorm on how to solve this new need, or we can of course do it in the Issue - basically we need an RFC with someone writing the initial design.

**Comment by stas00** [comment]
> hi @stas00 I have published the feature request at #3098.

@uygnef, this is the wrong destination. It needs to be moved to: https://github.com/huggingface/accelerate/issues

Diffusers is a user of Accelerate, it can't change Accelerate.

**Comment by tjruwase** [comment]
> Bottom line, do you want to meet and brainstorm on how to solve this new need, or we can of course do it in an Issue - basically we need an RFC with someone writing the initial design.

If it makes sense, I would love to be a part of such a brainstorming effort.

**Comment by pacman100** [comment]
Hello @stas00  and @tjruwase,  

> If I sound a bit bullish, it is not because I want to diminish these challenges, but rather because I am excited to be tackling these difficult yet unavoidable problems to support the next generation of ML models.

Completely agree with this!

> Bottom line, do you want to meet and brainstorm on how to solve this new need, or we can of course do it in an Issue - basically we need an RFC with someone writing the initial design.

I can take this up. Will try to put together a doc with the initial design and raise an RFC issue. ETA for this is Friday (21 April)

> If it makes sense, I would love to be a part of such a brainstorming effort.

Yes, that would be great, we can even have a quick sync up post having an initial design. Thank you!

**Comment by stas00** [comment]
That sounds like a plan, Sourab. Thank you for working on it!

**Comment by patrickvonplaten** [comment]
Thanks for the fixes here @stas00 ! 

This looks good to me! @patil-suraj @sayakpaul @williamberman could you also take a look?

File: examples/text_to_image/train_text_to_image.py [hunk 1]
```diff
@@ -461,10 +461,43 @@ def main():
     tokenizer = CLIPTokenizer.from_pretrained(
         args.pretrained_model_name_or_path, subfolder="tokenizer", revision=args.revision
     )
-    text_encoder = CLIPTextModel.from_pretrained(
-        args.pretrained_model_name_or_path, subfolder="text_encoder", revision=args.revision
-    )
-    vae = AutoencoderKL.from_pretrained(args.pretrained_model_name_or_path, subfolder="vae", revision=args.revision)
+
+    from accelerate.state import AcceleratorState
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image.py:465]: can we move those imports to the top?

File: examples/text_to_image/train_text_to_image.py [hunk 2]
```diff
@@ -461,10 +461,43 @@ def main():
...
+    # Currently Accelerate doesn't know how to handle multiple models under Deepspeed ZeRO stage 3.
+    # For this to work properly all models must be run through `accelerate.prepare`. But accelerate
+    # will try to assign the same optimizer with the same weights to all models during
+    # `deepspeed.initialize`, which of course doesn't work.
+    #
+    # For now the following workaround will partially support Deepspeed ZeRO-3, by excluding the 2
+    # frozen models from being partitioned during `zero.Init` which gets called during
+    # `from_pretrained` So CLIPTextModel and AutoencoderKL will not enjoy the parameter sharding
+    # across multiple gpus and only UNet2DConditionModel will get ZeRO sharded.
+    with ContextManagers(deepspeed_zero_init_disabled_context_manager()):
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image.py:493]: works for me!

File: src/diffusers/training_utils.py [hunk 3]
```diff
@@ -5,6 +5,7 @@
 
 import numpy as np
 import torch
+import transformers
```
Comment (patrickvonplaten) [inline src/diffusers/training_utils.py:8]: Actually `transformers` is still not a hard dependency (some training scripts don't need `transformers`)

File: examples/text_to_image/train_text_to_image.py [hunk 4]
```diff
@@ -461,10 +461,43 @@ def main():
     tokenizer = CLIPTokenizer.from_pretrained(
         args.pretrained_model_name_or_path, subfolder="tokenizer", revision=args.revision
     )
-    text_encoder = CLIPTextModel.from_pretrained(
-        args.pretrained_model_name_or_path, subfolder="text_encoder", revision=args.revision
-    )
-    vae = AutoencoderKL.from_pretrained(args.pretrained_model_name_or_path, subfolder="vae", revision=args.revision)
+
+    from accelerate.state import AcceleratorState
+    from transformers.utils import ContextManagers
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image.py:466]: here we need transformers so ok for me

File: src/diffusers/training_utils.py [hunk 5]
```diff
@@ -197,11 +198,17 @@ def step(self, parameters: Iterable[torch.nn.Parameter]):
...
+            import deepspeed
+        context_manager = []
         for s_param, param in zip(self.shadow_params, parameters):
-            if param.requires_grad:
-                s_param.sub_(one_minus_decay * (s_param - param))
-            else:
-                s_param.copy_(param)
+            if transformers.deepspeed.is_deepspeed_zero3_enabled():
+                context_manager = [deepspeed.zero.GatheredParameters(param, modifier_rank=None)]
+            with transformers.utils.ContextManagers(context_manager):
```
Comment (patrickvonplaten) [inline src/diffusers/training_utils.py:207]:
```suggestion
            context_manager = deepspeed.zero.GatheredParameters(param, modifier_rank=None) if is_deepspeed_enabled else contextlib.nullcontext
            with context_manager:
```

**Comment by patrickvonplaten** [comment]
In general, this looks good to me? Would be nice to not force-add `transformers` in `training_utils.py` as it's not a hard dependency. Ideally we can just have some nice `deepspeed` tools also in `diffusers`? Also cc @patil-suraj

**Comment by sayakpaul** [comment]
The solution looks minimal, clean, and good to me. More so, because we're aware of the limitations too (thanks @stas00 for that). Also thanks to @pacman100 and @tjruwase for their inputs and being open to resolving the current challenges :) 

I don't have additional comments on top of what @patrickvonplaten already shared.

**Comment by patrickvonplaten** [comment]
Tests are passing, let's merge :-)
